ledger/src/blocktree.rs

```diff
@@ -1182,25 +1186,75 @@ impl Blocktree {
         }
     }
 
-    // The `get_block_time` method is not fully implemented (depends on validator timestamp
-    // transactions). It currently returns Some(`slot` * DEFAULT_MS_PER_SLOT) offset from 0 for all
-    // transactions, and None for any values that would overflow any step.
-    pub fn get_block_time(&self, slot: Slot, slot_duration: Duration) -> Option<UnixTimestamp> {
-        let ms_per_slot = duration_as_ms(&slot_duration);
-        let (offset_millis, overflow) = slot.overflowing_mul(ms_per_slot);
-        if !overflow {
-            i64::try_from(offset_millis)
-                .ok()
-                .and_then(|millis| {
-                    let median_datetime = Utc.timestamp(0, 0);
-                    median_datetime.checked_add_signed(ChronoDuration::milliseconds(millis))
-                })
-                .map(|dt| dt.timestamp())
+    pub fn get_block_time(
+        &self,
+        slot: Slot,
+        slot_duration: Duration,
+        stakes: &HashMap<Pubkey, (u64, Account)>,
+    ) -> Option<UnixTimestamp> {
+        let mut total_stake = 0;
+        let stake_weighted_timestamps_sum: u64 = self
+            .get_timestamp_slots(slot, TIMESTAMP_SLOT_INTERVAL)
+            .iter()
+            .flat_map(|timestamp_slot| {
+                let offset = (slot - timestamp_slot) as u32 * slot_duration;
+                if let Ok(timestamps) = self.get_block_timestamps(*timestamp_slot) {
+                    timestamps
+                        .iter()
+                        .filter_map(|(vote_pubkey, timestamp)| {
+                            stakes.get(vote_pubkey).map(|(stake, _account)| {
+                                total_stake += stake;
+                                (*timestamp as u64 + offset.as_secs()) * stake
+                            })
+                        })
+                        .collect()
+                } else {
+                    vec![]
+                }
+            })
+            .sum();
+        if total_stake > 0 {
+            let mean_timestamp: u64 = stake_weighted_timestamps_sum / total_stake;
+            Some(mean_timestamp as i64)
         } else {
             None
         }
     }
 
+    fn get_timestamp_slots(&self, slot: Slot, timestamp_interval: u64) -> Vec<Slot> {
+        if self.is_root(slot) {
+            if let Ok(iter) = RootedSlotIterator::new(0, &self) {
```

Comment:
nit:
```suggestion
            if let Ok(rooted_slots) = RootedSlotIterator::new(0, &self) {
```

Calling this variable `iter` causes me brains to also match `iter()` as a scan the code, and I need to think harder than I should

Reply:
Fixed in db418ff